### PR TITLE
Fix nodes upload

### PR DESCRIPTION
**Pull request by aapomm**
Fixes: http://discuss.dradisframework.org/t/error-while-importing-owasp-project/492

With the new parent validation for nodes, the parsing of nodes in the template upload breaks because it relies on the fact that you can save nodes with invalid parent ID. This solution skips the validation when saving child nodes and only validate them properly when they have the right parent id.

### How to test:
In CE,
1. Download the OWASP package https://dradisframework.com/academy/industry/compliance/owasp/
2. Upload the file .zip file in the package
3. Assert that there were no errors in uploading.

### CODE REVIEW

**Comment by rachkor** [comment]
@aapomm I get the following error when I use this branch along with the `master` branch of `dradis-ce`: 
```
[17:09:02] Wrapping up...
[17:09:02] Setting issue_id for evidence
[17:09:02] Malformed Evidence detected: ["Node is invalid"]
[17:09:02] Couldn't save Evidence :issue_id / attachment URL Evidence #
[17:09:02] ["/Users/rachkor/Desktop/local-install/dradis-projects/lib/dradis/plugins/projects/upload/v1/template.rb:96:in `block in finalize_evidence'", <TRUNCATED>
[17:09:02] Done.
[17:09:02] Moving attachments to their final destinations...
[17:09:02] undefined method `[]' for false:FalseClass
[17:09:02] Worker process completed.
```

**Comment by aapomm** [comment]
@rachkor fixed. Can you check this again if this works for you?

File: lib/dradis/plugins/projects/upload/v1/template.rb [hunk 1]
```diff
@@ -218,28 +218,36 @@ def parse_node(xml_node)
...
-                .find_or_create_by!(label: label)
-          elsif [Node::Types::DEFAULT, Node::Types::HOST].exclude?(type_id.to_i)
-            Node.create_with(label: label)
-                .find_or_create_by!(type_id: type_id)
-          else
-            Node.create!(
+        if label == Configuration.plugin_uploads_node
+          node = Node.create_with(type_id: type_id, parent_id: parent_id)
+              .find_or_create_by!(label: label)
+        elsif [Node::Types::DEFAULT, Node::Types::HOST].exclude?(type_id.to_i)
```
Comment (dormi) [inline lib/dradis/plugins/projects/upload/v1/template.rb]: I've a pending change here: can we use `Node::Types::USER_HOSTS` please?

**Comment by dormi** [comment]
@aapomm I can only see 1 node created (`Uploaded files`). 
All imported evidence seems to be attached to that node.
Is that correct? 
uploaded_nodes

**Comment by rachkor** [comment]
Sorry @aapomm, I missed this notification! Feel free to ping me on Basecamp/Slack if this happens. 

The upload error has disappeared but the Node structure isn't being maintained. Here's what the OWASP project looks like on Pro (as intended): 

screen shot 2017-10-23 at 2 59 20 pm

But, in CE, all the Nodes are being added at the top level (I couldn't fit them all on the same screenshot but you can see that the `AUTHN` Nodes aren't nested as expected): 

screen shot 2017-10-23 at 2 59 58 pm

**Comment by aapomm** [comment]
@rachkor fixed the node structure!

**Comment by rachkor** [comment]
@aapomm the Node structure is looking great! I was doing some final checks and found that projects import just fine, but no longer export in a format that can be imported again. Can you try the steps to replicate? 

1. Import the OWASP sample project as `Package`
2. Export the resulting project as `Package` (`dradis-export.zip`)
3. Run `$ bundle exec thor dradis:reset`
4. Import `dradis-export.zip` from Step #2 as `Package` and note the following error message: 
```
[00:30:40] Processing Tags...
[00:30:40] New tag detected: !9467bd_critical
[00:30:40] Validation failed: Taggable can't be blank
[00:30:40] ["/Users/rachkor/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/activerecord-5.1.4/lib/active_record/validations.rb:78:in `raise_validation_error'", ... [truncated]
[00:30:40] Done.
[00:30:40] Moving attachments to their final destinations...
[00:30:40] undefined method `[]' for false:FalseClass
[00:30:40] Worker process completed.
```

Could you look into this and let me know?

**Comment by aapomm** [comment]
@rachkor upon inspection, it would seem that this error exists in the default setup and not because of this PR. I'll look into this more.

**Comment by rachkor** [comment]
@aapomm Tested and confirmed the fix with the latest version of `dradis-ce` 👍

**Pull request merged**